File: homework.py
import time
from copy import copy, deepcopy

class Cell(object):

	#def __init__(self, player, value):
		#self.player = player
		#self.value = value
	def __init__(self,value):
		self.value = value
		self.mymove = ""

	def markplayer(self, player):
		self.player = player
	def indexi(self, i):
		self.indexi = i
	def indexj(self, j):
		self.indexj = j

def alphabeta(state, player):

	v, action = max_value_alpha(state, 0, player, float("-inf"), float("inf"))
	printSolBoard(action)
	print(v)

def max_value_alpha(state, depth, player, alpha, beta):

	if terminal_test(state) or depth == max_depth: 
		return utility(state)
# generate a new board with new child
	v = alpha
	best_b = None
	list_act = actions(state, player)
	oplayer = otherPlayer(player)
	for b in list_act:
		v , best_b = max_func(v, min_value_alpha(b, depth + 1, oplayer, v, beta), best_b, b)
		if v >= beta: 
			return v, best_b
		alpha = max(alpha, v)
	return v, b
def min_value_alpha(state, depth, player, alpha, beta):
	if terminal_test(state) or depth == max_depth: 
		return utility(state)
	v = beta
	best_b = None
	list_act = actions(state, player)
	oplayer = otherPlayer(player)
	for b in list_act:
		v, best_b= min_func(v, max_value_alpha(b, depth + 1, oplayer, alpha, v), best_b, b)
		if v <= alpha: 
			return v, best_b
		beta = min(beta, v)
	return v, b

def minimax(state, player):
    v, action = max_value(state, 0, player)
    printSolBoard(action)
    print(v)
    #return v, action
def max_value(state, depth, player):
	if terminal_test(state) or depth == max_depth: 
		return utility(state)
# generate a new board with new child
	v = float("-inf")
	best_b = None
	list_act = actions(state, player)
	oplayer = otherPlayer(player)
	for b in list_act:
		v , best_b = max_func(v, min_value(b, depth + 1, oplayer), best_b, b)
		#printBoard(len(b), b)
	return v, best_b

def min_value(state, depth, player):
	if terminal_test(state) or depth == max_depth: 
		return utility(state)
	v = float("inf")
	best_b = None
	list_act = actions(state, player)
	oplayer = otherPlayer(player)
	for b in list_act:
		v,best_b= min_func(v, max_value(b, depth + 1, oplayer), best_b, b)
	#printBoard(len(b), b)
	return v, best_b 

def max_func(prev, curr, best_b, b):
	if type(curr) is tuple:
		curr = curr[0]
	if prev > curr:
		return prev, best_b
	elif curr > prev:
		return curr,b
	else:
		#check that stake is preffered over raid
		if best_b != None and b != None:
			tup = check(best_b, b)
			if tup == 1:
				return prev, best_b
			else:
				return curr, b
		else:
			return prev, best_b

def min_func(prev, curr, best_b, b):
	if type(curr) is tuple:
		curr = curr[0]
	if prev < curr:
		return prev, best_b
	elif curr < prev:
		return curr, b
	else:
		#check that stake is preffered over raid
		if best_b != None and b != None:
			tup = check(best_b, b)
			if tup == 1:
				return prev, best_b
			else:
				return curr, b
		else:
			return prev, best_b

def check(bestb, b):
	for i in range(len(bestb)):
		for j in range(len(bestb)):
			cell = bestb[i][j]
			if cell.mymove == "Steak":
				return 1
	for l in range(len(b)):
		for m in range(len(b)):
			cell = b[l][m]
			if cell.mymove == "Steak":
				return 2

def utility(state):
	maxInt = 0
	oplayer = otherPlayer(main_player)
	for i in range(len(state)):
		for j in range(len(state)):
			cell = state[i][j]
			if cell.player == main_player:
				maxInt += cell.value
			if cell.player == oplayer:
				maxInt -= cell.value			
	return maxInt

def stake(player, board, i, j):
	cell = board[i][j]
	if cell.player == '.':
		newBoard = deepcopy(board)
		new_cell = newBoard[i][j]
		new_cell.markplayer(player)
		new_cell.mymove = "Stake"
		new_cell.indexi(i)
		new_cell.indexj(j)
		#printBoard(length, newBoard)
		return newBoard
	return None

def raid(player, board, i, j):
	length = len(board)
	oplayer = otherPlayer(player)
	cell = board[i][j]
	newBoardList = []
	if cell.player == player:
		#left of occupied piece
		if j - 1 >= 0 and board[i][j - 1].player == '.':
			newBoard = deepcopy(board)
			new_cell = newBoard[i][j - 1]
			new_cell.markplayer(player)
			new_cell.mymove = "Raid"
			new_cell.indexi(i)
			new_cell.indexj(j - 1)

			if j - 2 >= 0 and newBoard[i][j - 2].player == oplayer: 
				newBoard[i][j - 2].markplayer(player)
			if i + 1 <= length - 1 and j - 1 >= 0 and newBoard[i + 1][j - 1].player == oplayer:
				newBoard[i + 1][j - 1].markplayer(player)
			if i - 1 >= 0 and j - 1 <= length - 1 and newBoard[i - 1][j - 1].player == oplayer:
				newBoard[i - 1][j - 1].markplayer(player)
			#printBoard(length, newBoard)
			newBoardList.append(newBoard)
		#right of occupied piece
		if j + 1 <= length - 1 and board[i][j + 1].player == '.':
			newBoard = deepcopy(board)
			new_cell = newBoard[i][j + 1]
			new_cell.markplayer(player)
			new_cell.mymove = "Raid"
			new_cell.indexi(i)
			new_cell.indexj(j + 1)

			if j + 2 <= length - 1 and newBoard[i][j + 2].player == oplayer:
				newBoard[i][j + 2].markplayer(player)
			if i + 1 <= length - 1 and j + 1 <= length and newBoard[i + 1][j + 1].player == oplayer:
				newBoard[i + 1][j + 1].markplayer(player)
			if i - 1 >= 0 and j + 1 <= length - 1 and newBoard[i - 1][j + 1].player == oplayer:
				newBoard[i - 1][j + 1].markplayer(player)
			#printBoard(length, newBoard)
			newBoardList.append(newBoard)
		#below of occupied piece
		if i + 1 <= length - 1 and board[i + 1][j].player == '.':
			newBoard = deepcopy(board)
			new_cell = newBoard[i + 1][j]
			new_cell.markplayer(player)
			new_cell.mymove = "Raid"
			new_cell.indexi(i + 1)
			new_cell.indexj(j)

			if i + 1 <= length - 1 and j - 1 >= 0 and newBoard[i + 1][j -1].player == oplayer:
				newBoard[i + 1][j - 1].markplayer(player)
			if i + 1 <= length - 1 and j + 1 <= length - 1 and newBoard[i + 1][j + 1].player == oplayer:
				newBoard[i + 1][j + 1].markplayer(player)
			if i + 2 <= length - 1 and newBoard[i + 2][j].player == oplayer:
				newBoard[i + 2][j].markplayer(player)
			#printBoard(length, newBoard)
			newBoardList.append(newBoard)
		#above occupied piece
		if i - 1 >= 0 and board[i - 1][j].player == '.':
			newBoard = deepcopy(board)
			new_cell = newBoard[i - 1][j]
			new_cell.markplayer(player)
			new_cell.mymove= "Raid"
			new_cell.indexi(i - 1)
			new_cell.indexj(j)

			if i - 1 >= 0 and j - 1 >= 0 and newBoard[i - 1][j - 1].player == oplayer:
				newBoard[i - 1][j - 1].markplayer(player)
			if i - 1 >= 0 and j + 1 <= length - 1 and newBoard[i - 1][j + 1].player == oplayer:
				newBoard[i - 1][j + 1].markplayer(player)
			if i - 2 >= 0 and newBoard[i - 2][j].player == oplayer:
				newBoard[i - 2][j].markplayer(player)
			#printBoard(length, newBoard)
			newBoardList.append(newBoard)
	#Did not find a valid move
	if newBoardList:
		return newBoardList
	else:
		return None
def terminal_test(state):
	#return boolean if we are at determined depth or all the spaces are filled
	length = len(state)
	for i in range(length):
		for j in range(length):
			#cell = 
			if state[i][j].player == '.':
				return False
	return True

def otherPlayer(player):
	otherplayer = ''
	if player == 'X':
		otherplayer = 'O'
	else:
		otherplayer = 'X'
	return otherplayer

def actions(state, player):
	length = len(state)
	moves = []
	for i in range(length):
		for j in range(length):
			stakeboard = stake(player, state, i, j)
			raidboardList = raid(player, state, i, j)

			if stakeboard != None:
				moves.append(stakeboard)
			if raidboardList != None:
				moves.extend(raidboardList)
	if moves:
		return moves
	else:
	 	return None

def printSolBoard(board):
	f = open('output.txt', 'w')
	length = len(board)
	row = 0
	col = ''
	move = ''
	for i in range(length):
		for j in range(length):
			cell = board[i][j]
			if cell.mymove != "":
				move = cell.mymove
				row = cell.indexi + 1
				col = chr(cell.indexj + ord('A'))
				f.write(col)
				f.write(str(row) + " ")
				f.write(move)
				f.write('\n')

	for k in range(length):
		for l in range(length):		
			cell = board[k][l]
			f.write(cell.player)
		f.write('\n')
	f.close()

def printBoard(n, board):
	for i in range(n):
		for j in range(n):
			c = board[i][j]
			print(c.value, end='')
			print(c.player, end=' ')
		print()

def main():
	with open('input.txt', 'r') as f:
		n = int((f.readline().split())[0])
		mode = f.readline().split()[0]
		global main_player
		main_player = f.readline().split()[0]
		global max_depth 
		max_depth = int((f.readline().split())[0])
		board = [[0 for x in range(n)] for y in range(n)]
		for i in range(n):
			l = f.readline().split()
			for j in range(n):
				board[i][j] = Cell(int(l[j]))
		for m in range(n):
			lin = list(f.readline().split()[0])
			for k in range(n):
				cell = board[m][k]
				cell.markplayer(lin[k])
	f.close()
	if mode == "MINIMAX":
	 	minimax(board, main_player)
	elif mode == "ALPHABETA":
	 	alphabeta(board, main_player)


if __name__ == "__main__":
	main()

File: homework.java
/*
* Author: Michelle Becerra
* Date: 10/17/16
* Foundations of Artificial Intelligence Homework 2 Minimax and Alpha Beta pruning
* Professor Dr. Itti
*/

import java.util.*;
import java.io.*;

class homework {

	//GLOBAL VARIABLES
	private static char main_player = Character.MIN_VALUE;
	private static int max_depth = 0;
	private static HashMap<Integer,Cell[][]> boardMap = new HashMap<Integer, Cell[][]>();
	
	//Calls Parse File which parses the file and determines which algorithms needs to run
	public static void main(String[] args) throws FileNotFoundException {

		parseFile();
	}

	public static void parseFile() throws FileNotFoundException {

		Scanner br = new Scanner(new File("input.txt"));
		// Size of board
		int n = 0;
		if (br.hasNextInt()) {
			n = br.nextInt();
		}
		// Mode of algorithm
		String mode = "";
		if (br.hasNextLine()) {
			mode = br.next();
		}
		// Player
		if (br.hasNextLine()) {
			main_player = br.next().charAt(0);
		}
		// Max depth
		if (br.hasNextInt()) {
			max_depth = br.nextInt();
		}

		// Creates nxn matrix of cell objects
		Cell[][] board = new Cell[n][n];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				int val = br.nextInt();
				board[i][j] = new Cell(val);
			}
		}

		// Adds players to the cell objects in nxn matrix
		for (int i = 0; i < n; i++) {
			String line = br.next();
			for (int j = 0; j < n; j++) {
				char c = line.charAt(j);
				if ('X' == c || 'O' == c) {
					board[i][j].setPlayer(c);
				}
			}
		}

		//Process either minimax or alpha beta pruning
		if (mode.equals("MINIMAX")) {

			minimax(board);
		}
		else if (mode.equals("ALPHABETA")) {
			alphabeta(board);
		}
		else {
			br.close();
			return;
		}

		br.close();
	}
	//Alpha beta pruning - Outputs to output.txt 
	public static void alphabeta(Cell[][] board) throws FileNotFoundException{
		int val;
		val = max_value_alpha(board, 0, Integer.MIN_VALUE, Integer.MAX_VALUE);
		printSolBoard(val, board.length);
	}
	public static int max_value_alpha(Cell[][] board, int depth, int alpha, int beta) {
		if(terminal_test(board) || depth == max_depth){
			return utility(board);
		}

		
		int length = board.length;
		Cell[][] best_c = new Cell[length][length];
		List<Cell[][]>  successors = actions(board, main_player);

		Boolean root_level = false;
		if(depth == 0) root_level = true;

		int value = Integer.MIN_VALUE;
		//best_c = successors.get(0);
		int val;
		
		for(Cell[][] child : successors){
			val =  min_value_alpha(child, depth + 1, alpha, beta);
			if(val > value){
				best_c = child;
				value = val;
			}

			if( value >= beta){
				boardMap.put(value, best_c);
				return value;
			}
			alpha = Math.max(alpha, value);
		}

		boardMap.put(value, best_c);
		return value;
	}
	public static int min_value_alpha(Cell[][] board, int depth, int alpha, int beta){
		if(terminal_test(board) || depth == max_depth){
			return utility(board);
		}
		int value = Integer.MAX_VALUE;
		char oplayer = otherPlayer(main_player);
		List<Cell[][]>  successors = actions(board, oplayer);
		int length = board.length;
		Cell[][] best_c = new Cell[length][length];
		int val;
		for(Cell[][] child: successors){
			val = max_value_alpha(child, depth + 1, alpha, beta);
			if( val < value){
				best_c = child;
				value = val;
			}

			if(value <= alpha){
				boardMap.put(value, best_c);
				return value;
			}
			beta = Math.min(beta, value);
		}
		boardMap.put(value, best_c);
		return value;
	}
	//Minimax - Outputs to output.txt 
	public static void minimax(Cell[][] board) throws FileNotFoundException {
		int val;
		val = max_value(board, 0);
		printSolBoard(val, board.length);
	}
	public static int max_value(Cell[][] board, int depth) {
		if(terminal_test(board) || depth == max_depth){
			return utility(board);
		}
		int value = Integer.MIN_VALUE;
		List<Cell[][]>  successors = actions(board, main_player);
		int length = board.length;
		Cell[][] best_c = new Cell[length][length];
		//best_c = successors.get(0);
		Boolean root_level = false;
		if(depth == 0) root_level = true;
		int val;
		for(Cell[][] child: successors){
			val = min_value(child, depth + 1);
			
			if( val > value){
				best_c = child;
				value = val;
			}			
		}
		//create a hashmap mapping value to board
		boardMap.put(value, best_c);
		return value;
	}
	public static int min_value(Cell[][] board, int depth){
		if(terminal_test(board) || depth == max_depth){
			return utility(board);
		}
		int value = Integer.MAX_VALUE;
		char oplayer = otherPlayer(main_player);
		List<Cell[][]>  successors = actions(board, oplayer);
		int length = board.length;
		Cell[][] best_c = new Cell[length][length];
		//best_c = successors.get(0);
		int val;
		for(Cell[][] child: successors){
			val = max_value(child, depth + 1);
			if( val < value){
				best_c = child;
				value = val;
			}
		}
		boardMap.put(value, best_c);		
		return value;
	}

	public static int utility(Cell[][] board) {
		int maxInt = 0;

		char oplayer = otherPlayer(main_player);
		
		int length = board.length;
		for (int i = 0; i < length; i++) {
			for (int j = 0; j < length; j++) {
				Cell cell = board[i][j];

				if(cell.player() == main_player){
					maxInt += cell.value();
				}
				if (cell.player() == oplayer) {
					maxInt -= cell.value();
				}

			}
		}
		return maxInt;
	}
	//Gets the possible actions for current player
	public static List<Cell[][]> actions(Cell[][] board, char player){
		int length = board.length;
		//Process stake moves first
		List<Cell[][]> moves = new ArrayList<Cell[][]>();
		for(int i = 0; i < length; i ++){
			for(int j = 0; j < length; j++){
				Cell[][] stakeBoard = stake(board,player, i, j);
				if(stakeBoard != null){
					moves.add(stakeBoard);
				}

			}
		}
		//Process raid moves next
		for(int l = 0; l < length; l++){
			for(int m = 0; m < length; m++){
				List<Cell[][]> raidList = raidNew(board,player, l, m);
				if(raidList != null){
					for(Cell[][] raidBoard: raidList){
						moves.add(raidBoard);
					}
				}

			}
		}
		return moves;
	}
	//Stake move - can move to an unocuppied space only
	public static Cell[][] stake(Cell[][] board, char player, int i, int j) {
		 Cell cell = board[i][j];
		 if (cell.player() == '.') {
			Cell[][] newBoard = deepCopy(board);
			Cell new_cell = newBoard[i][j];
			new_cell.setPlayer(player);
			new_cell.setMove("Stake");
			new_cell.setIndexi(i);
			new_cell.setIndexj(j);
			//printBoard(newBoard);
			return newBoard;
		 }
		 return null;
	}

	public static List<Cell[][]> raidNew(Cell[][] board, char player, int i, int j){
		Cell cell = board[i][j];
		int length = board.length;
		char oplayer = otherPlayer(player);
		List<Cell[][]> boardList = new ArrayList<Cell[][]>();

		 if (cell.player() == '.') {
		 	
		 	//Player is at left
		 	if (j - 1 >= 0 && board[i][j - 1].player() == player){
				Cell[][] newBoard1 = deepCopy(board);
				newBoard1[i][j].setMove("Raid");
				newBoard1[i][j].setIndexi(i);
				newBoard1[i][j].setIndexj(j);
				newBoard1[i][j].setPlayer(player);
	
				//Check if we can conquer adjacent pieces

				//right
				if (j + 1 <= length - 1 && newBoard1[i][j + 1].player() == oplayer){
					newBoard1[i][j + 1].setPlayer(player);
				}
				//down 
				if (i + 1 <= length - 1 && newBoard1[i + 1][j].player() == oplayer){
					newBoard1[i + 1][j].setPlayer(player);
				}
				//up
				if (i - 1 >= 0 && newBoard1[i - 1][j].player() == oplayer){
					newBoard1[i - 1][j].setPlayer(player);
				}
				boardList.add(newBoard1);	
				//return boardList;
		 	}
		 	//PLayer is at right
		 	else if(j + 1 <= length - 1 && board[i][j + 1].player() == player){
		 		Cell[][] newBoard = deepCopy(board);
		 		newBoard[i][j].setMove("Raid");
		 		newBoard[i][j].setIndexi(i);
				newBoard[i][j].setIndexj(j);
				newBoard[i][j].setPlayer(player);

				//Check if we can conquer adjacent pieces
				//left
				if(j - 1 >= 0 && newBoard[i][j - 1].player() == oplayer){
					newBoard[i][j - 1].setPlayer(player);
				} 
				//down
				if(i + 1 <= length - 1 && newBoard[i + 1][j].player() == oplayer){
					newBoard[i + 1][j].setPlayer(player);
				}
				//up
				if(i - 1 >= 0 && newBoard[i - 1][j].player() == oplayer){
					newBoard[i - 1][j].setPlayer(player);
				}
				boardList.add(newBoard);
				//return boardList;
		 	}
		 	//Player is below me
		 	else if(i + 1 <= length - 1 && board[i + 1][j].player() == player){
		 		Cell[][] newBoard3 = deepCopy(board);
				newBoard3[i][j].setMove("Raid");
				newBoard3[i][j].setIndexi(i);
				newBoard3[i][j].setIndexj(j);
				newBoard3[i][j].setPlayer(player);
	
				//left
				if (j - 1 >= 0 && newBoard3[i][j - 1].player() == oplayer){
					newBoard3[i][j - 1].setPlayer(player);
				}
				//right
				if (j + 1 <= length - 1 && newBoard3[i][j + 1].player() == oplayer){
					newBoard3[i][j + 1].setPlayer(player);
				}
				//above
				if (i - 1 >= 0 && newBoard3[i - 1][j].player() == oplayer){
					newBoard3[i - 1][j].setPlayer(player);
				}
				boardList.add(newBoard3);
				//return boardList;
				
		 	}
		 	//Player is above me
		 	else if(i - 1 >= 0 && board[i - 1][j].player() == player){
		 		Cell[][] newBoard2 = deepCopy(board);
		 		newBoard2[i][j].setMove("Raid");
		 		newBoard2[i][j].setIndexi(i);
				newBoard2[i][j].setIndexj(j);
				newBoard2[i][j].setPlayer(player);

				//Check if we can conquer adjacent pieces
				//left
				if (j - 1 >= 0 && newBoard2[i][j -1].player() == oplayer){
					newBoard2[i][j - 1].setPlayer(player);
				}
				//right
				if (j + 1 <= length - 1 && newBoard2[i][j + 1].player() == oplayer){
					newBoard2[i][j + 1].setPlayer(player);
				}
				//below
				if (i + 1 <= length - 1 && newBoard2[i + 1][j].player() == oplayer){
					newBoard2[i + 1][j].setPlayer(player);
				}
				boardList.add(newBoard2);
				//return boardList;
		 	}
		 	return boardList;
		 }
		 return null;
	}
	//Raid move - can take over an unoccupied space with another player in the same
	//            team occupies a neighboring piece. Can then conquer adjacent pieces
	public static List<Cell[][]> raid(Cell[][] board, char player,  int i, int j) {
		char oplayer = otherPlayer(player);
		Cell cell = board[i][j];
		int length = board.length;
		List<Cell[][]> boardList = new ArrayList<Cell[][]>();

	//CHECK that a player on the same team is to the Left, Right, Down or Up 
	if (cell.player() == player){
		
		//Left of occupied piece
		if (j - 1 >= 0 && board[i][j - 1].player() == '.'){
			Cell[][] newBoard = deepCopy(board);
			Cell new_cell = newBoard[i][j - 1];
			new_cell.setPlayer(player);
			new_cell.setMove("Raid");
			new_cell.setIndexi(i);
			new_cell.setIndexj(j - 1);

			//Check if we can conquer adjacent pieces
			if(j - 2 >= 0 && newBoard[i][j - 2].player() == oplayer){
				newBoard[i][j - 2].setPlayer(player);
			} 
			if(i + 1 <= length - 1 && j - 1 >= 0 && newBoard[i + 1][j - 1].player() == oplayer){
				newBoard[i + 1][j - 1].setPlayer(player);
			}
			if(i - 1 >= 0 && j - 1 <= length - 1 && newBoard[i - 1][j - 1].player() == oplayer){
				newBoard[i - 1][j - 1].setPlayer(player);
			}
			boardList.add(newBoard);
		}
		//Right of occupied piece
		if (j + 1 <= length - 1 && board[i][j + 1].player() == '.'){
			Cell[][] newBoard1 = deepCopy(board);
			Cell new_cell1 = newBoard1[i][j + 1];
			new_cell1.setPlayer(player);
			new_cell1.setMove("Raid");
			new_cell1.setIndexi(i);
			new_cell1.setIndexj(j + 1);

			//Check if we can conquer adjacent pieces
			if (j + 2 <= length - 1 && newBoard1[i][j + 2].player() == oplayer){
				newBoard1[i][j + 2].setPlayer(player);
			}
			if (i + 1 <= length - 1 && j + 1 <= length && newBoard1[i + 1][j + 1].player() == oplayer){
				newBoard1[i + 1][j + 1].setPlayer(player);
			}
			if (i - 1 >= 0 && j + 1 <= length - 1 && newBoard1[i - 1][j + 1].player() == oplayer){
				newBoard1[i - 1][j + 1].setPlayer(player);
			}
			boardList.add(newBoard1);
		}
		//Below of occupied piece
		if (i + 1 <= length - 1 && board[i + 1][j].player() == '.'){
			Cell[][] newBoard2 = deepCopy(board);
			Cell new_cell2 = newBoard2[i + 1][j];
			new_cell2.setPlayer(player);
			new_cell2.setMove("Raid");
			new_cell2.setIndexi(i + 1);
			new_cell2.setIndexj(j);

			//Check if we can conquer adjacent pieces
			if (i + 1 <= length - 1 && j - 1 >= 0 && newBoard2[i + 1][j -1].player() == oplayer){
				newBoard2[i + 1][j - 1].setPlayer(player);
			}
			if (i + 1 <= length - 1 && j + 1 <= length - 1 && newBoard2[i + 1][j + 1].player() == oplayer){
				newBoard2[i + 1][j + 1].setPlayer(player);
			}	
			if (i + 2 <= length - 1 && newBoard2[i + 2][j].player() == oplayer){
				newBoard2[i + 2][j].setPlayer(player);
			}
			boardList.add(newBoard2);
		}
		//Above occupied piece
		if (i - 1 >= 0 && board[i - 1][j].player() == '.'){
			Cell[][] newBoard3 = deepCopy(board);
			Cell new_cell3 = newBoard3[i - 1][j];
			new_cell3.setPlayer(player);
			new_cell3.setMove("Raid");
			new_cell3.setIndexi(i - 1);
			new_cell3.setIndexj(j);

			if (i - 1 >= 0 && j - 1 >= 0 && newBoard3[i - 1][j - 1].player() == oplayer){
				newBoard3[i - 1][j - 1].setPlayer(player);
			}
			if (i - 1 >= 0 && j + 1 <= length - 1 && newBoard3[i - 1][j + 1].player() == oplayer){
				newBoard3[i - 1][j + 1].setPlayer(player);
			}
			if (i - 2 >= 0 && newBoard3[i - 2][j].player() == oplayer){
				newBoard3[i - 2][j].setPlayer(player);
			}
			boardList.add(newBoard3);
		}
		return boardList;
	}
		return null;

	}
	//Checks to see if there any unoccupied spaces
	public static boolean terminal_test(Cell[][] board) {
		int length = board.length;
		for (int i = 0; i < length; i++) {
			for (int j = 0; j < length; j++) {
				Cell cell = board[i][j];

				if (cell.player() == '.') {
					return false;
				}
			}
		}
		return true;
	}
	//Returns opposite player
	public static char otherPlayer(char player) {

		if ('X' == player) {
			return 'O';
		}
		else {
			return 'X';
		}
	}
	//Prints the solution move to output.txt
	public static void printSolBoard(int value, int length) throws FileNotFoundException {
		PrintWriter printer = new PrintWriter(new File("output.txt"));
		Cell[][] board = boardMap.get(value);
		String move = "";
		String col = "";
		int row;
		for (int i = 0; i < length; i++) {
			for (int j = 0; j < length; j++) {
				Cell cell = board[i][j];
				//Ouput move first in format: [A-Z][1-26] "Steak/Raid"
				if (!cell.move().equals("")) {
					move = cell.move();
					row = cell.indexi() + 1;
					col = String.valueOf((char)(cell.indexj() + 'A'));
					printer.write(col);
					printer.write(row + " ");
					printer.write(move);
					printer.write("\n");

				}
			}
		}
		//Output the pieces/players of the solution board
		for (int k = 0; k < length; k++) {
			for (int l = 0; l < length; l++) {
				Cell cell = board[k][l];
				printer.write(cell.player());
			}
			printer.write("\n");
		}
		printer.close();
	}
	//Creates a deep copy of the board and a deep copy of the cell objets
	public static Cell[][] deepCopy(Cell[][] old){
		Cell[][] newBoard = new Cell[old.length][old.length];
		
		for(int i = 0; i < old.length; i++){
			for(int j = 0; j < old.length; j++){
				newBoard[i][j] = new Cell(old[i][j]);
			}
		}
		return newBoard;
	}
	//Created for my testing purposes
	public static void printBoard(Cell[][] board) {
		int length = board.length;
		for (int i = 0; i < length; i++) {
			for (int j = 0; j < length; j++) {
				Cell cell = board[i][j];

				System.out.print(cell.value() +""+ cell.player() + " ");	
			}
		    System.out.println();
		}
	}

}


